cpp/src/phonenumbers/phonenumberutil.cc

```diff
@@ -1527,12 +1525,76 @@ void PhoneNumberUtil::FormatNsnWithCarrier(const string& number,
       : metadata.intl_number_format();
   const NumberFormat* formatting_pattern =
       ChooseFormattingPatternForNumber(available_formats, number);
+  NumberFormat new_format;
+  bool consistent_optional_rule = TRUE;
+  bool consistent_fg_rule = TRUE;
```

Comment:
Similarly here - you can track `formatting_rule_has_first_group_only` (or something less verbose).  Defaults to `FALSE` but if any format says `TRUE`, then set it to `TRUE`.